internal/testrunner/runners/system/servicedeployer/custom_agent.go

```diff
@@ -0,0 +1,198 @@
+	opts := compose.CommandOptions{
+		Env:       env,
+		ExtraArgs: []string{"--build", "-d"},
+	}
+	err = p.Up(opts)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not boot up service using Docker Compose")
+	}
+
+	// Connect service network with stack network (for the purpose of metrics collection)
+	err = docker.ConnectToNetwork(p.ContainerName(serviceName), stack.Network())
+	if err != nil {
+		return nil, errors.Wrapf(err, "can't attach service container to the stack network")
+	}
+
+	err = p.WaitForHealthy(opts)
+	if err != nil {
+		processServiceContainerLogs(p, compose.CommandOptions{
+			Env: opts.Env,
+		}, outCtxt.Name)
+		return nil, errors.Wrap(err, "service is unhealthy")
+	}
+
+	// Build service container name
+	outCtxt.Hostname = p.ContainerName(serviceName)
+
+	logger.Debugf("adding service container %s internal ports to context", p.ContainerName(serviceName))
+	serviceComposeConfig, err := p.Config(compose.CommandOptions{Env: env})
+	if err != nil {
+		return nil, errors.Wrap(err, "could not get Docker Compose configuration for service")
+	}
+
+	s := serviceComposeConfig.Services[serviceName]
+	outCtxt.Ports = make([]int, len(s.Ports))
+	for idx, port := range s.Ports {
+		outCtxt.Ports[idx] = port.InternalPort
+	}
+
+	// Shortcut to first port for convenience
+	if len(outCtxt.Ports) > 0 {
+		outCtxt.Port = outCtxt.Ports[0]
+	}
+
+	outCtxt.Agent.Host.NamePrefix = inCtxt.Name
+	service.ctxt = outCtxt
+	return &deployedCustomAgent{
+		dockerComposeDeployedService: &service,
+		cfg:                          d.cfg,
+		env:                          env,
+	}, nil
+}
+
+// TearDown tears down the service.
+func (s *deployedCustomAgent) TearDown() error {
+	defer func() {
+		if err := os.Remove(s.cfg); err != nil {
+			logger.Errorf("cleaning up tmp file (path: %s): %v", s.cfg, err)
+		}
+	}()
+	return s.dockerComposeDeployedService.TearDown()
+}
+
+func createCustomAgentYaml(cfgPath string) (string, error) {
+	bc := struct {
+		Version  string                            `yaml:"version"`
+		Services map[string]map[string]interface{} `yaml:"services"`
+	}{}
+	if err := yaml.Unmarshal([]byte(customAgentYml), &bc); err != nil {
+		return "", errors.Wrap(err, "unmarshal base custom-agent config")
+	}
+
+	cb, err := os.ReadFile(cfgPath)
+	if err != nil {
+		return "", errors.Wrap(err, "open custom-agent config")
+	}
+
+	cv := map[string]interface{}{}
+	if err := yaml.Unmarshal(cb, &cv); err != nil {
+		return "", errors.Wrap(err, "unmarshal custom-agent config")
+	}
+
+	for k, v := range cv {
+		bc.Services[dockerCustomAgentName][k] = v
+	}
+	bc.Services[dockerCustomAgentName]["hostname"] = dockerCustomAgentName
+
+	b, err := yaml.Marshal(bc)
+	if err != nil {
+		return "", errors.Wrap(err, "marshal custom-agent config")
+	}
+
+	tf, err := os.CreateTemp("", dockerCustomAgentName)
+	if err != nil {
+		return "", errors.Wrap(err, "create tmp file")
+	}
+	defer tf.Close()
+
+	if _, err := tf.Write(b); err != nil {
```

Comment:
You don't need to use the temp dir if you "install" `custom-agent.yml` to the `~/.elastic-package/deployer/custom-agent` similarly to `~/.elastic-package/deployer/terraform/terraform-deployer.yml`. This would be a very similar mechanism for extending Compose definition.